lib/elixir/lib/module/checker.ex

```diff
@@ -0,0 +1,135 @@
+defmodule Module.Checker do
+  def verify(module_map, _binary) do
+    check_definitions(Enum.reverse(module_map.definitions), module_map)
+    collect_warnings()
+  end
+
+  defp collect_warnings() do
+    receive do
+      {:warning, file, line, message} ->
+        [{file, line, message} | collect_warnings()]
+    after
+      0 ->
+        []
+    end
+  end
+
+  defp check_definitions(definitions, state) do
+    Enum.each(definitions, &check_definition(&1, state))
+  end
+
+  defp check_definition({def, _kind, meta, clauses}, state) do
```

Comment:
We should build our own state instead of changing the `module_map`. It is best to be clear about what we need and what we don't. Ideally the `module_map` is read-only.